web/src/ResourceGroupsContext.test.tsx

```diff
@@ -0,0 +1,152 @@
+import { mount, ReactWrapper } from "enzyme"
+import React from "react"
+import { AnalyticsAction, AnalyticsType } from "./analytics"
+import {
+  cleanupMockAnalyticsCalls,
+  expectIncrs,
+  mockAnalyticsCalls,
+} from "./analytics_test_helpers"
+import {
+  DEFAULT_GROUP_STATE,
+  ResourceGroupsContextProvider,
+  useResourceGroups,
+} from "./ResourceGroupsContext"
+
+const GROUP_STATE_ID = "test-group-state"
+const LABEL_STATE_ID = "test-label-state"
+
+// This is a very basic test component that prints out the state
+// from the ResourceGroups context and provides buttons to trigger
+// methods returned by the context, so they can be tested
+const TestConsumer = (props: { labelName?: string }) => {
+  const { groups, getGroup, toggleGroupExpanded } = useResourceGroups()
+
+  return (
+    <>
+      <p id={GROUP_STATE_ID}>{JSON.stringify(groups)}</p>
+      {/* Display the label state if a specific label is present */}
+      {props.labelName && (
+        <p id={LABEL_STATE_ID}>{JSON.stringify(getGroup(props.labelName))}</p>
+      )}
+      {/* Display a button to toggle the label state if a specific label is present */}
+      {props.labelName && (
+        <button
+          onClick={() =>
+            toggleGroupExpanded(props.labelName || "", AnalyticsType.Grid)
+          }
+        />
+      )}
+    </>
+  )
+}
```

Comment:
i'm not sure if this is the best way to test a provider/context, but looking around our codebase and (very briefly) the internet, it seems like this is a common approach. open to suggestions if there's a better way to test!

Reply:
this seems like a reasonable approach. we've also done this:
https://github.com/tilt-dev/tilt/blob/master/web/src/ResourceNav.test.tsx#L15
which uses the "Consumer" react component (rather than hooks) and attaches the state to a class (rather than serializing/deserializing). but i don't think either approach is necessarily more correct.

i might add a helper method for the stanza:

```
      const labelState = wrapper.find(`#${LABEL_STATE_ID}`).text()
      expect(labelState).toBe(JSON.stringify({ expanded: true }))
```

Reply:
ah, cool! it's good to see another example for testing. added a couple small helpers!